snuba/web/query.py

```diff
@@ -234,9 +262,41 @@ def log_query_and_update_stats(
     return stats
 
 
-@split_query
 def parse_and_run_query(
     dataset: Dataset, request: Request, timer: Timer
+) -> ClickhouseQueryResult:
+    """
+    Runs a query, then records the results including metadata about each split.
+    """
+    query_list: List[ClickhouseQueryMetadata] = []
+    try:
+        result = _run_query(
+            dataset=dataset, request=request, timer=timer, query_list=query_list
+        )
+    except RawQueryException as error:
+        raise error
+    finally:
+        if settings.RECORD_QUERIES:
+            # send to redis
+            state.record_query(
+                {
+                    "request": request.body,
+                    "referrer": http_request.referrer,
+                    "timing": query_list[-1].timer,
+                    "status": query_list[-1].status,
+                    "query_list": [q.to_dict() for q in query_list],
```

Comment:
I haven't found a great solution for this, but there is now an implicit assumption that all of the values returned as part of `ClickhouseQueryMetadata.to_dict`, are JSON serializable.

I think this is the case (and it is at least protected by a try/except if they aren't) but I don't think this gets exercised by tests since `RECORD_QUERIES` defaults to `False`.